hub/bolt_transport.go

```diff
@@ -0,0 +1,239 @@
+}
+
+// Write pushes updates in the Transport
+func (t *BoltTransport) Write(update *Update) error {
+	select {
+	case <-t.done:
+		return ErrClosedTransport
+	default:
+	}
+
+	var (
+		err         error
+		closedPipes []*Pipe
+	)
+
+	t.RLock()
+
+	err = t.persist(update)
+
+	for pipe := range t.pipes {
+		if !pipe.Write(update) {
+			closedPipes = append(closedPipes, pipe)
+		}
+	}
+
+	t.RUnlock()
+	t.Lock()
+
+	for _, pipe := range closedPipes {
+		delete(t.pipes, pipe)
+	}
+
+	t.Unlock()
+
+	return err
+}
+
+// persist stores update in the database
+func (t *BoltTransport) persist(update *Update) error {
+	return t.db.Update(func(tx *bolt.Tx) error {
+		bucket, err := tx.CreateBucketIfNotExists([]byte(t.bucketName))
+		if err != nil {
+			return err
+		}
+
+		buf, err := json.Marshal(*update)
+		if err != nil {
+			return err
+		}
+
+		seq, err := bucket.NextSequence()
+		if err != nil {
+			return err
+		}
+		t.lastSeq.Store(seq)
+		prefix := make([]byte, 8)
+		binary.BigEndian.PutUint64(prefix, seq)
+
+		// The sequence value is prepended to the update id to create an ordered list
+		key := bytes.Join([][]byte{prefix, []byte(update.ID)}, []byte{})
+
+		if err := t.cleanup(bucket, seq); err != nil {
+			return err
+		}
+
+		// The DB is append only
+		bucket.FillPercent = 1
+		return bucket.Put(key, buf)
+	})
+}
+
+// CreatePipe returns a pipe fetching updates from the given point in time
+func (t *BoltTransport) CreatePipe(fromID string) (*Pipe, error) {
+	t.Lock()
+	defer t.Unlock()
+
+	select {
+	case <-t.done:
+		return nil, ErrClosedTransport
+	default:
+	}
+
+	pipe := NewPipe()
+	t.pipes[pipe] = struct{}{}
+	if fromID == "" {
+		return pipe, nil
+	}
+
+	toSeq := t.lastSeq.Load()
+	go func() {
+		err := t.db.View(func(tx *bolt.Tx) error {
+			b := tx.Bucket([]byte(t.bucketName))
+			if b == nil {
+				return nil // No data
+			}
+
+			c := b.Cursor()
+			afterFromID := false
+			for k, v := c.First(); k != nil; k, v = c.Next() {
+				if !afterFromID {
+					if string(k[8:]) == fromID {
+						afterFromID = true
+					}
+
+					continue
+				}
+
+				var update *Update
+				if err := json.Unmarshal(v, &update); err != nil {
+					return err
+				}
+
+				if !pipe.Write(update) || (toSeq > 0 && binary.BigEndian.Uint64(k[:8]) >= toSeq) {
+					return nil
+				}
+			}
+
+			return nil
+		})
+		if err != nil {
+			log.Error(fmt.Errorf("bolt history: %w", err))
+		}
+	}()
+
+	return pipe, nil
+}
+
+// Close closes the Transport
+func (t *BoltTransport) Close() error {
+	select {
+	case <-t.done:
+		// Already closed. Don't close again.
+	default:
+		t.RLock()
+		defer t.RUnlock()
+		for pipe := range t.pipes {
+			pipe.Close()
+		}
+		close(t.done)
+		t.db.Close()
+	}
+
+	return nil
+}
+
+// cleanup removes entries in the history above the size limit, triggered probabilistically
+func (t *BoltTransport) cleanup(bucket *bolt.Bucket, lastID uint64) error {
```

Comment:
I wonder if that method shouldn't run in a goroutine called every x seconds

Reply:
why not, but maybe in another PR?